ui/src/main/java/org/pentaho/di/ui/trans/step/BaseStepDialog.java

```diff
@@ -807,6 +808,10 @@ String showDbDialogUnlessCancelledOrValid( DatabaseMeta changing, DatabaseMeta o
     cid.setDatabaseMeta( changing );
     cid.setModalDialog( true );
 
+    if ( cid.getDatabaseMeta() == null ) {
```

Comment:
What is the relation of this change and the actual bug being fixed?

Reply:
The logic here is to check if the databaseMeta is not null.  With the implementation of the implementation of Security on Datasources this would be set to null if the user didn't have the correct permissions.  This allows the dialog to abort (but not before the system displays the correct error).  Again this does't fix a BUG this is a new FEATURE.